File: app/src/main/java/com/veganbeauty/app/data/local/UserMemoryManager.java
package com.veganbeauty.app.data.local;

import android.content.Context;

import com.veganbeauty.app.data.local.entities.YtVideoEntity;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class UserMemoryManager {

    public static class HandbookCategory {
        private String id;
        private String name;
        private List<YtVideoEntity> videos;

        public HandbookCategory(String id, String name, List<YtVideoEntity> videos) {
            this.id = id != null ? id : UUID.randomUUID().toString();
            this.name = name;
            this.videos = videos != null ? videos : new ArrayList<>();
        }

        public String getId() { return id; }
        public String getName() { return name; }
        public List<YtVideoEntity> getVideos() { return videos; }
        
        public void setName(String name) { this.name = name; }
    }

    private final Context context;
    private final String fileName = "user_handbook_memory.json";
    private static final String LEGACY_FILE_NAME = "user_memory.json";
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public UserMemoryManager(Context context) {
        this.context = context;
    }

    public List<HandbookCategory> getCategories() {
        File file = resolveHandbookFile();
        if (!file.exists()) return new ArrayList<>();
        List<HandbookCategory> list = new ArrayList<>();
        try (FileInputStream fis = new FileInputStream(file);
             InputStreamReader isr = new InputStreamReader(fis, StandardCharsets.UTF_8);
             BufferedReader reader = new BufferedReader(isr)) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            JSONArray array = new JSONArray(sb.toString());
            for (int i = 0; i < array.length(); i++) {
                JSONObject obj = array.getJSONObject(i);
                HandbookCategory cat = new HandbookCategory(
                        obj.optString("id", UUID.randomUUID().toString()),
                        obj.optString("name", ""),
                        new ArrayList<>()
                );
                JSONArray videosArr = obj.optJSONArray("videos");
                if (videosArr != null) {
                    for (int j = 0; j < videosArr.length(); j++) {
                        JSONObject vObj = videosArr.getJSONObject(j);
                        cat.getVideos().add(new YtVideoEntity(
                                vObj.optString("id", ""),
                                vObj.optString("title", ""),
                                vObj.optString("url", ""),
                                vObj.optString("description", ""),
                                vObj.optString("username", ""),
                                vObj.has("avatarUrl") && !vObj.isNull("avatarUrl") ? vObj.optString("avatarUrl") : null,
                                vObj.optString("type", "")
                        ));
                    }
                }
                list.add(cat);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return list;
    }

    private File resolveHandbookFile() {
        File file = new File(context.getFilesDir(), fileName);
        if (file.exists()) {
            return file;
        }
        File legacy = new File(context.getFilesDir(), LEGACY_FILE_NAME);
        if (legacy.exists()) {
            return legacy;
        }
        return file;
    }

    public void saveCategories(List<HandbookCategory> categories) {
        File file = new File(context.getFilesDir(), fileName);
        try {
            JSONArray array = new JSONArray();
            for (HandbookCategory cat : categories) {
                JSONObject obj = new JSONObject();
                obj.put("id", cat.getId());
                obj.put("name", cat.getName());
                JSONArray videosArr = new JSONArray();
                for (YtVideoEntity v : cat.getVideos()) {
                    JSONObject vObj = new JSONObject();
                    vObj.put("id", v.getId());
                    vObj.put("title", v.getTitle());
                    vObj.put("url", v.getUrl());
                    vObj.put("description", v.getDescription());
                    vObj.put("username", v.getUsername());
                    vObj.put("avatarUrl", v.getAvatarUrl());
                    vObj.put("type", v.getType());
                    videosArr.put(vObj);
                }
                obj.put("videos", videosArr);
                array.put(obj);
            }
            String jsonString = array.toString();
            try (FileOutputStream fos = new FileOutputStream(file)) {
                fos.write(jsonString.getBytes(StandardCharsets.UTF_8));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void addVideoToCategory(String categoryName, YtVideoEntity video) {
        List<HandbookCategory> categories = getCategories();
        HandbookCategory category = null;
        for (HandbookCategory cat : categories) {
            if (cat.getName().equals(categoryName)) {
                category = cat;
                break;
            }
        }
        if (category == null) {
            category = new HandbookCategory(null, categoryName, null);
            categories.add(category);
        }
        
        boolean exists = false;
        for (YtVideoEntity v : category.getVideos()) {
            if (v.getId().equals(video.getId()) || v.getUrl().equals(video.getUrl())) {
                exists = true;
                break;
            }
        }
        if (!exists) {
            category.getVideos().add(video);
            saveCategories(categories);
        }
    }

    public void removeVideoFromCategory(String categoryName, YtVideoEntity video) {
        List<HandbookCategory> categories = getCategories();
        for (HandbookCategory category : categories) {
            if (category.getName().equals(categoryName)) {
                for (int i = 0; i < category.getVideos().size(); i++) {
                    YtVideoEntity v = category.getVideos().get(i);
                    if (v.getId().equals(video.getId()) || v.getUrl().equals(video.getUrl())) {
                        category.getVideos().remove(i);
                        break;
                    }
                }
                saveCategories(categories);
                break;
            }
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/data/local/UserMemoryManager.kt
package com.veganbeauty.app.data.local

import android.content.Context
import com.veganbeauty.app.data.local.entities.YtVideoEntity
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

data class HandbookCategory(
    val id: String = java.util.UUID.randomUUID().toString(),
    var name: String,
    val videos: MutableList<YtVideoEntity> = mutableListOf()
)

class UserMemoryManager(private val context: Context) {
    private val fileName = "user_memory.json"

    fun getCategories(): MutableList<HandbookCategory> {
        val file = File(context.filesDir, fileName)
        if (!file.exists()) return mutableListOf()
        val list = mutableListOf<HandbookCategory>()
        try {
            val jsonString = file.readText()
            val array = JSONArray(jsonString)
            for (i in 0 until array.length()) {
                val obj = array.getJSONObject(i)
                val cat = HandbookCategory(
                    id = obj.optString("id", java.util.UUID.randomUUID().toString()),
                    name = obj.optString("name", "")
                )
                val videosArr = obj.optJSONArray("videos")
                if (videosArr != null) {
                    for (j in 0 until videosArr.length()) {
                        val vObj = videosArr.getJSONObject(j)
                        cat.videos.add(
                            YtVideoEntity(
                                id = vObj.optString("id", ""),
                                title = vObj.optString("title", ""),
                                url = vObj.optString("url", ""),
                                description = vObj.optString("description", ""),
                                username = vObj.optString("username", ""),
                                avatarUrl = if (vObj.has("avatarUrl") && !vObj.isNull("avatarUrl")) vObj.optString("avatarUrl") else null,
                                type = vObj.optString("type", "")
                            )
                        )
                    }
                }
                list.add(cat)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return list
    }

    fun saveCategories(categories: List<HandbookCategory>) {
        val file = File(context.filesDir, fileName)
        try {
            val array = JSONArray()
            for (cat in categories) {
                val obj = JSONObject()
                obj.put("id", cat.id)
                obj.put("name", cat.name)
                val videosArr = JSONArray()
                for (v in cat.videos) {
                    val vObj = JSONObject()
                    vObj.put("id", v.id)
                    vObj.put("title", v.title)
                    vObj.put("url", v.url)
                    vObj.put("description", v.description)
                    vObj.put("username", v.username)
                    vObj.put("avatarUrl", v.avatarUrl)
                    vObj.put("type", v.type)
                    videosArr.put(vObj)
                }
                obj.put("videos", videosArr)
                array.put(obj)
            }
            val jsonString = array.toString()
            file.writeText(jsonString)
            
            // Sync to Firebase background
            kotlinx.coroutines.GlobalScope.launch(kotlinx.coroutines.Dispatchers.IO) {
                try {
                    val db = com.google.firebase.firestore.FirebaseFirestore.getInstance()
                    val data = hashMapOf("handbookData" to jsonString)
                    db.collection("user_handbooks").document("test_001").set(data).await()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun addVideoToCategory(categoryName: String, video: YtVideoEntity) {
        val categories = getCategories()
        var category = categories.find { it.name == categoryName }
        if (category == null) {
            category = HandbookCategory(name = categoryName)
            categories.add(category)
        }
        if (category.videos.none { it.id == video.id || it.url == video.url }) {
            category.videos.add(video)
            saveCategories(categories)
        }
    }

    fun removeVideoFromCategory(categoryName: String, video: YtVideoEntity) {
        val categories = getCategories()
        val category = categories.find { it.name == categoryName }
        if (category != null) {
            category.videos.removeAll { it.id == video.id || it.url == video.url }
            saveCategories(categories)
        }
    }
}
